File: core/SubtreeOfAnotherTree/SubtreeOfAnotherTree.java
package com.dbc.code;

public class SubtreeOfAnotherTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private boolean res = false;
    private String dfs(TreeNode node, String flag) {
        if (node == null) return "null";
        String leftNode = dfs(node.left, flag);
        String rightNode = dfs(node.right, flag);
        String nodeStr = String.format("l%s|%d|r%s", leftNode, node.val, rightNode);
        if (nodeStr.equals(flag)) {
            this.res = true;
        }
        return nodeStr;
    }

    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        String subStr = dfs(subRoot, null);
        dfs(root, subStr);
        return this.res;
    }
}

File: core/SubtreeOfAnotherTree/SubtreeOfAnotherTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def isSubtree(root: TreeNode, subRoot: TreeNode) -> bool:
    res = False

    def dfs(node: TreeNode, flag: str):
        nonlocal res
        if not node:
            return "None"
        left_str = dfs(node.left, flag)
        right_str = dfs(node.right, flag)
        node_str = f"l{left_str}|{node.val}|r{right_str}"
        if flag and node_str == flag:
            res = True
        return node_str

    sub_str = dfs(subRoot, None)
    dfs(root, sub_str)
    return res
